### PR TITLE
Update attribute() signature and add/update some docstrings.

### PR DESCRIPTION
**Pull request by Funth0mas**
As promissed, I unified the signature of attribute() functions.
Added docstrings to canmatrix.py.
Fixed few typos.
Add some remarks (comments ending with question mark)

Sorry to mix again these change types but it was single-pass work. I still believe it's reviewable, I strongly talked to my fingers not to do _all_ they wanted to :-) . I hope it helps to increase the code lucidity.

Thx for review.

### CODE REVIEW

File: src/canmatrix/xls_common.py [hunk 1]
```diff
@@ -84,17 +84,17 @@ def getSignal(db, sig, motorolaBitFormat):
     # startvalue of signal available
     if "GenSigStartValue" in db.signalDefines:
         if db.signalDefines["GenSigStartValue"].definition == "STRING":
-            frontArray.append(sig.attribute(db,"GenSigStartValue"))
+            frontArray.append(sig.attribute("GenSigStartValue"), db=db)
```
Comment (altendky) [inline src/canmatrix/xls_common.py]: Need to shift a paren outside the `db=db.  Presently it has `db` as a parameter to `.append()`.

File: src/canmatrix/canmatrix.py [hunk 2]
```diff
@@ -94,15 +106,18 @@ def normalizeValueTable(table):
...
+    Signal has following attributes:
+
+    * name
+    * startBit, size (in Bits)
+    * is_little_endian (1: Intel, 0: Motorola)
+    * is_signed (bool)
+    * factor, offset, min, max
+    * receiver  (Boarunit/ECU-Name)
+    * attributes, _values, unit, comment
+    * _multiplex ('Multiplexor' or Number of Multiplex)
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: Is this a standard doc form?  I can't say I've done much with docs but if we are filling this stuff in it might be worth finding doc generation that can work with attrs.

File: src/canmatrix/canmatrix.py [hunk 3]
```diff
@@ -338,27 +396,40 @@ def raw2phys(self, value, decodeToStr=False):
...
+
 class SignalGroup(object):
     """
-    contains Signals, which belong to signal-group
+    Represents signal-group, containing multiple Signals.
     """
 
     def __init__(self, name, Id):
-        self.signals = []
+        self.signals = []  # use attr.ib?
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: Separate PR, but sure, go for it.

File: src/canmatrix/canmatrix.py [hunk 4]
```diff
@@ -338,27 +396,40 @@ def raw2phys(self, value, decodeToStr=False):
...
+        """Remove Signal from SignalGroup.
+
+        :param Signal signal: signal to remove
+        """
         if signal in self.signals:
-            self.signals[signal].remove()
+            self.signals.remove(signal)
 
-    def byName(self, name):
+    def byName(self, name):  # add 'Signal' to function name?
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: It's a signal group, what would you get other than signals?  I mean, maybe, but that's my gut response.  :]

File: src/canmatrix/canmatrix.py [hunk 5]
```diff
@@ -517,10 +596,12 @@ def addSignalGroup(self, Name, Id, signalNames):
             if signalId is not None:
                 newGroup.addSignal(signalId)
 
-    def signalGroupbyName(self, name):
+    def signalGroupbyName(self, name):  # capital B: ByName?
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: I would say yes.

Though I would `snake_case` anyways.  That'd be a separate issue I presume.

File: src/canmatrix/canmatrix.py [hunk 6]
```diff
@@ -853,8 +951,8 @@ def saveConvertStrToInt(inStr):
         elif definition[0:3] == 'HEX':
             self.type = 'HEX'
             min, max = definition[4:].split(' ', 2)
-            self.min = saveConvertStrToInt(min)
-            self.max = saveConvertStrToInt(max)
+            self.min = safeConvertStrToInt(min)  # the same for INT and HEX?
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: I agree, that doesn't sound right.  But, I don't know offhand what triggers this code.

File: src/canmatrix/canmatrix.py [hunk 7]
```diff
@@ -1119,11 +1287,18 @@ def recalcDLC(self, strategy):
...
-        if type(old).__name__ == 'instance':
+        """Rename ECU in the Matrix. Update references in all Frames.
+
+        :param str or BoardUnit old: old name or ECU instance
+        :param str newName: new name
+        """
+        if type(old).__name__ == 'instance':  # accept BoardUnit instance
             pass
-        else:
+        else:  # or string
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: Yeah, we need to stop doing this type guessing stuff.  I'd say either `renameEcuByBoardUnit()` and `renameEcuByName()` or we just check if old is in `self.boardUnits` and not then treat it as a string.  But, this is a separate issue.

Not to mention that when we do check type it should be `isinstance(this, That)`. :]  Looks like there are a few places to apply this.

File: src/canmatrix/canmatrix.py [hunk 8]
```diff
@@ -1227,43 +1421,77 @@ def delSignal(self, signal):
...
+        :param BoardUnit ecu: Receiver ECU name
+        """
+        frames = self.globFrames(framePattern)
         for frame in frames:
             signal = frame.signalByName(signalName)
             if signal is not None:
                 signal.addReceiver(ecu)
                 for signal in frame.globSignals(signalName):
-                    signal.addReceiver(ecu)
+                    signal.addReceiver(ecu)  # already set, redundant line
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: Huh.  This seems very explicit but yeah, I'm not sure the cases where `frame.globSignals(signalName)` would get additional signals.  @ebroecker?  See similar below as well.

File: src/canmatrix/canmatrix.py [hunk 9]
```diff
@@ -1278,9 +1506,10 @@ def merge(self, mergeArray):
                         "Name Conflict, could not copy/merge EnvVar " + envVar)
 
     def setFdType(self):
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: This sounds more like `guessIfFdAndSet()` or somesuch.  @ebroecker, what's the history on this?  We can't tell if a frame is FD or not?  Or can you not have an FD frame with <= 8 bytes?  Why only setting to `True` and no case for setting `False`?

File: src/canmatrix/canmatrix.py (hunk 3 above)
Comment (ebroecker) [reply src/canmatrix/canmatrix.py]: Oh ... YES. I just forgot SignalGroup while switching to attr.ib

File: src/canmatrix/canmatrix.py (hunk 7 above)
Comment (ebroecker) [reply src/canmatrix/canmatrix.py]: I fully agree

File: src/canmatrix/canmatrix.py [hunk 10]
```diff
@@ -1136,24 +1311,26 @@ def renameEcu(self, old, newName):
             frame.updateReceiver()
 
     def addEcu(self, ecu):
+        """Add new ECU to the Matrix. Do nothing if ecu with the same name already exists.
```
Comment (ebroecker) [inline src/canmatrix/canmatrix.py]: I mixed up Ecu and BU (=BoardUnit) --
One name should be enough.
Which one we should choose?

File: src/canmatrix/canmatrix.py [hunk 11]
```diff
@@ -1174,8 +1352,13 @@ def updateEcuList(self):
...
 
     def renameFrame(self, old, newName):
+        """Rename Frame.
+
+        :param Frame or str old: Old Frame instance or name or part of the name with '*' at the beginning or the end.
+        :param str newName: new Frame name, suffix or prefix
+        """
         if type(old).__name__ == 'instance':
-            oldName = old.name
+            old = old.name
```
Comment (ebroecker) [inline src/canmatrix/canmatrix.py]: + you found a bug!

File: src/canmatrix/canmatrix.py (hunk 8 above)
Comment (ebroecker) [reply src/canmatrix/canmatrix.py]: I could imagine, that the indentation is wrong here.
The Idea was to check first if there is a signal with the exact name.
Afterwards check if there is a signal searching with globSignals.

But this makes no sense.
There should only be the "globSignals" way to find the right signals, the "signalByName" could be removed?!

File: src/canmatrix/canmatrix.py [hunk 12]
```diff
@@ -1227,43 +1421,77 @@ def delSignal(self, signal):
...
+        """
+        frames = self.globFrames(framePattern)
         for frame in frames:
             signal = frame.signalByName(signalName)
             if signal is not None:
-                signal.addReceiver(ecu)
+                signal.delReceiver(ecu)
                 for signal in frame.globSignals(signalName):
-                    signal.delReceiver(ecu)
+                    signal.delReceiver(ecu)  # redundant
```
Comment (ebroecker) [inline src/canmatrix/canmatrix.py]: I think same copy paste issue from addSignalReceiver

File: src/canmatrix/canmatrix.py (hunk 9 above)
Comment (ebroecker) [reply src/canmatrix/canmatrix.py]: you are right - this is only guessing.
Meanwhile with J1939 this may even do mistakes.
In deed I think FD has long DLCs only - but I may be wrong. If I am wrong, setting False would be wrong also :-)

File: src/canmatrix/xls_common.py (hunk 1 above)
Comment (ebroecker) [reply src/canmatrix/xls_common.py]: You are right, this must be fixed

File: src/canmatrix/canmatrix.py (hunk 2 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: Yes. Try ctrl+Q in pycharm, it gets rendered.

Comment (altendky) [reply src/canmatrix/canmatrix.py]: I use pycharm, but it wouldn't be my primary target when picking a doc format.

File: src/canmatrix/canmatrix.py (hunk 5 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: It might be hard job, the refactoring capabilities of pycharm aren't unlimited as I saw :-( But yes, the fingers want to write snake case.

File: src/canmatrix/canmatrix.py (hunk 8 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: Mainly I did not remarked the 2nd `addReceiver` runs on another object with the same name. I'll try to thing about it once againe.

File: src/canmatrix/canmatrix.py (hunk 2 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: :-) No idea but I hope the sphinx can be converted to another format "automatically"

File: src/canmatrix/canmatrix.py (hunk 10 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: Personally I've never heard 'Board Unit' but always 'ECU'. But it doesn't matter. The abbreviation 'ECU' seems clearer than 'BU' for me. You decide :-)

File: src/canmatrix/canmatrix.py (hunk 9 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: I thing the data speed depends merely on the flag in the CAN Frame header on the bus. That means "DLC>8 implies CAN FD" but DLC <= 8 implies nothing.

The function could than rather be `checkFdRequired()`. And set back to False, if there is nothing with long DLC. Agree?

File: src/canmatrix/canmatrix.py (hunk 4 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: But than, in code, you see `return tmp.byName(name)` and you're lost... The __getitem__() might be aslo useful. Postpone to another PR.. maybe

File: src/canmatrix/canmatrix.py (hunk 7 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: Leaving the code untouched for further refactoring.

File: src/canmatrix/canmatrix.py (hunk 10 above)
Comment (ebroecker) [reply src/canmatrix/canmatrix.py]: OK, ECU should be the name for it in future.

File: src/canmatrix/canmatrix.py (hunk 9 above)
Comment (ebroecker) [reply src/canmatrix/canmatrix.py]: Agree

**Comment by ebroecker** [comment]
is there anything open, or could we merge?

File: src/canmatrix/canmatrix.py (hunk 9 above)
Comment (Funth0mas) [reply src/canmatrix/canmatrix.py]: I'll propose it separately so that this "big" PR contains minimum changes.

File: src/canmatrix/canmatrix.py [hunk 13]
```diff
@@ -158,70 +174,93 @@ def multiplexSetter(self, value):
         if value is not None and value != 'Multiplexor':
             ret_multiplex = int(value)
             self.mux_val = int(value)
-        else:
+        else:  # is it valid for None too?
```
Comment (Funth0mas) [inline src/canmatrix/canmatrix.py]: this is still unclear but it might be resolved in the future

**Comment by Funth0mas** [comment]
I checked the diff once again and I don't see any issue which should be part of this PR. Rather for further work.

Which concerns the docstring format, they should be convertible by pyment.

File: src/canmatrix/canmatrix.py (hunk 4 above)
Comment (altendky) [reply src/canmatrix/canmatrix.py]: I wouldn't use method names to help callers keep track of what sort of object they have.  It seems like the name referring to the object itself is what should be sufficiently descriptive for the context.

File: src/canmatrix/canmatrix.py [hunk 14]
```diff
@@ -1227,43 +1417,71 @@ def delSignal(self, signal):
...
-    def addSignalReceiver(self, frameName, signalName, ecu):
-        frames = self.globFrames(frameName)
+    def addSignalReceiver(self, framePattern, signalName, ecu):
+        """Add Receiver to all Frames by name pattern.
+
+        :param str framePattern: Frame name pattern
+        :param str signalName: signal name
+        :param str ecu: Receiver ECU name
+        """
+        frames = self.globFrames(framePattern)
```
Comment (altendky) [inline src/canmatrix/canmatrix.py]: If it's going to be exclusively a glob I'd use that in the name rather than the generic 'pattern'.  Ditto for the docstring.  Glob, like regex, isn't a single thing but it's a _relatively_ standard pattern so I'd think it's useful to be explicit.

May as well catch `signalName` as well if we're going to exclusively glob it.  Also, find all similar below (or move them to their own PR).

File: src/canmatrix/canmatrix.py (hunk 9 above)
Comment (altendky) [reply src/canmatrix/canmatrix.py]: As is this could probably be at least 3 I think.  :]  Every new change that's added holds up all the rest from being accepted.

File: src/canmatrix/canmatrix.py (hunk 11 above)
Comment (altendky) [reply src/canmatrix/canmatrix.py]: What has a class name of (lower-case) `instance` anyways?

**Comment by Funth0mas** [comment]
The code with 'instance' is not mine. For the rest - please excuse my anarchism, I'll revert all issues you mentioned back to the previous version. I'll be more careful next time. Thank you for your patience.
